### PR TITLE
noexcept requirement ScopeGuard

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

This PR adds a static assertion that the lambda that is passed to the scope guard is noexcept. The old scope guard ignored all exceptions silently, possibly leaving the server in an unknown or invalid state. This was required because the lambda is executed in a destructor. With these changes the code actively requires the user to think about the exception case.

This PR annotates a lot of functions with noexcept that were not yet defined as such.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

- [x] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16835/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16838/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16839/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16841/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16844/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16845/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16847/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16853/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16862/

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 1]
```diff
@@ -1211,7 +1211,7 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
     });
   };
 
-  TRI_DEFER(shutdown());
+  auto sg = arangodb::scopeGuard([&]() noexcept { shutdown(); });
```
Comment (mpoeter) [inline arangod/Replication/DatabaseInitialSyncer.cpp:1214]: `shutdown` is not noexcept and can potentially throw AFAIS.

File: arangosh/Dump/DumpFeature.cpp [hunk 2]
```diff
@@ -777,7 +777,7 @@ Result DumpFeature::runSingleDump(httpclient::SimpleHttpClient& client,
   if (res.fail()) {
     return res;
   }
-  TRI_DEFER(::endBatch(client, "", batchId));
+  auto sg = arangodb::scopeGuard([&]() noexcept { ::endBatch(client, "", batchId); });
```
Comment (mpoeter) [inline arangosh/Dump/DumpFeature.cpp:780]: `endBatch` is not `noexcept` and can potentially throw AFAIS.

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16872/

**Pull request merged**